### ISSUE TITLE
hyperopt does not use all cores, and hyperopt gets slower

### ISSUE THREAD

**Issue by squat0001** [freqtrade/freqtrade#3258]
## Step 1: Have you search for this issue before posting it?

If you have discovered a bug in the bot, please [search our issue tracker](https://github.com/freqtrade/freqtrade/issues?q=is%3Aissue). 
If it hasn't been reported, please create a new issue.

## Step 2: Describe your environment

  * Operating system:  DEBIAN BUSTER
  * Python Version: 3.3  (`python -V`)
  * Branch: Master
  * Last Release 2020.4
  
## Step 3: Describe the problem:

hyperopt does not utilize all CPU cores, i tested some settings with joblib and cpu affinity, like
described in: (it only uses 1 cpu at default)
https://stackoverflow.com/questions/15639779/why-does-multiprocessing-use-only-a-single-core-after-i-import-numpy

and adding support for masking cpu cores does make joblib use them.. but hyperopt also does
get slower real fast.

attached is a small patch just to show that there is a difference in behavior and cpu usage.
it does (it does not fix the problem)
my test system has 4 cpu cores
- set the affinity for the master process
- set the affinity for the slave processes
- set the joblib verbose to see how long a job takes
- hack the parameter "-j" to "jobs" instead of cores to get a better yield

[fix.txt](https://github.com/freqtrade/freqtrade/files/4577192/fix.txt)

and here the verbose log of hyperopt joblib .. look at the times:

it starts with 4 secs per backtesting.. 
after about 8000 epochs

[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.                                                                                                        [1957/1957]
[Parallel(n_jobs=4)]: Done   5 tasks      | elapsed:   22.5s
[Parallel(n_jobs=4)]: Done  11 out of  16 | elapsed:   49.1s remaining:   22.3s
[Parallel(n_jobs=4)]: Done  13 out of  16 | elapsed:   57.9s remaining:   13.4s
[Parallel(n_jobs=4)]: Done  16 out of  16 | elapsed:  1.2min finished
.
.
.(about 1000 epochs later)
[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done   5 tasks      | elapsed:   33.6s
[Parallel(n_jobs=4)]: Done  11 out of  16 | elapsed:  1.2min remaining:   33.4s
[Parallel(n_jobs=4)]: Done  13 out of  16 | elapsed:  1.4min remaining:   20.0s
[Parallel(n_jobs=4)]: Done  16 out of  16 | elapsed:  1.8min finished

**Comment by hroff-1902** [comment]
did not get the point, to be honest...

yes, the behavior of hyperopt is suboptimal: when it runs N epochs simultaneously (where N is either set explicitly with `-j N` or fetched as the number of CPUs by default) in N joblib worker processes, it currently waits for the completion of all of them, i.e. waits for the longest of those N epochs,.. It's a known issue and we have a couple of long-awating PRs to change it.

But I did not get anything about 'affinity', 'numpy', 'jobs instead of cores' above... Could you describe this in more detail?

**Comment by xmatthias** [comment]
The fact that all jobs wait for "each other" is known, and we got a PR to fix this in #2951.

Now for the rest, i agree with @hroff-1902.

I would suggest to test the code from the PR above - it should make sure all cores are utilized.
Now to be fair, i didn't look at this PR in quite some time, so i'm not sure at what point it is, but i think it's pretty much complete.

**Comment by squat0001** [comment]
Hi,

Am 05.05.2020 um 06:10 schrieb Matthias:
>
> The fact that all jobs wait for "each other" is known, and we got a PR
> to fix this in #2951 <https://github.com/freqtrade/freqtrade/pull/2951>.
>
> Now for the rest, i agree with @hroff-1902
> <https://github.com/hroff-1902>.
>
> I would suggest to test the code from the PR above - it should make
> sure all cores are utilized.
> Now to be fair, i didn't look at this PR in quite some time, so i'm
> not sure at what point it is, but i think it's pretty much complete.
>
yes it looks much better than the current implementation which is in
many ways incomplete.

what do you need to merge it?

**Comment by hroff-1902** [comment]
> do you need to merge it?

An amount of spare time with good mood and concentration to review, test and merge... 😉

**Comment by xmatthias** [comment]
well we're all locked down, so the time should not be the issue.

The concentration though might be - as it's not just us beeing locked down, but also the family etc. ...

**Comment by squat0001** [comment]
Hi,

Am 10.05.2020 um 09:59 schrieb Matthias:
>
> well we're all locked down, so the time should not be the issue.
>
> The concentration though might be - as it's not just us beeing locked
> down, but also the family etc. ...
>
the merge did apply nearly without any conflict. the only thing which i
merged "away" was the progressbar.
(i used the head branch of his repository, not the pull request, because
he did many improvements to it)
(he only changes the way backtesting is called, backtesting is not touched)

the new hyperopt is about 5-10 times faster than the one in main stream,
and its optizming and not only probing by random.
(the current implementation takes about 20hours for 5000 epochs in my
example, the new takes about 20min for 5000 epochs)

the progressbar is hard to remerge, because the new implementation does
not need the parameter "epoch" it uses the optimizer to
search the complete space.

in the new version also the cpu upper bound of the optimizer is removed.
its possible to run the optimizer itself parallel, so for systems with
many cores
its possible to utilize them a lot more.

to get something like the progressbar .. i enabled the verbose mode of
joblib again.

greetings

**Comment by xmatthias** [comment]
I think you misunderstand why that PR is not merged.

Merge-conflicts are almost never a constraining factor - these can usually be solved one way or another.
The big problem is to ensure that all current features work, work the same way, and don't break on one or the other system or circumstance.

Just because it returns 5000 epochs in 20 mins does not necessarily mean that the results will be similar / comparable - maybe it did 5000 epochs, and 3500 of these were identical because randomness breaks after 500 epochs...

Now i don't say that's what happens, but it would be one possibility - therefore we're taking our time reviewing this PR to be sure hyperopt works the way we expect it to afterwards.

**Comment by devilsinthedetail** [comment]
@squat0001 How did you merge the head branch of his repository? I tried it but runs into conflicts with  " object has no attribute 'amounts'". I guess thats something new @orehunt added, and I only know its a dict but what it should contain is a bit puzzling.

**Comment by sdfsdsdsd** [comment]
@xmatthias @hroff-1902 greetings

i just realized that on a 32 core machine only one core is used and found this issue.

I fully understand that this takes time to merge, but this situation is quite sad, do you think there is a chance this happens anytime soon?

cheers and thanks for the great work

**Comment by xmatthias** [comment]
@sdfsdsdsd it's unlikely to be merged soon (if ever) as there's still a TON of work to do to get that PR even close to mergable - and it won't really give you what you expect either.

The PR uses disconnected processes - so it's like you're running 8 parallel hyperopts without any connection / communication - so while it may use 100% of your cpu, the results will most likely be worse (which is not a sacrifice i'm willing to make just to use 100% of the cpu).


The real problem isn't hyperopt itself, but the underlying skopt algorhithm, which is not parallelizable, so backtesting runs in whatever many processes in parallel - and then syncs (and waits) for the optimizer to generate the next point in one core (the base algorithm itself is apparently no parallelizable).

*edit*: adding source for the above: https://github.com/scikit-optimize/scikit-optimize/issues/768

**Comment by sdfsdsdsd** [comment]
@xmatthias thanks.... now i understand a bit better and i agree it doesnt seem to make sense to just decouple the hyperopts from each other.

for what it's worth i just found this https://github.com/scikit-optimize/scikit-optimize/issues/737

do you see hope in other tools or algorithms?
